assets/src/edit-story/components/panels/elementAlignment.js

```diff
@@ -0,0 +1,366 @@
+    const centerX = (boundRect.endX + boundRect.startX) / 2;
+    onSetProperties((properties) => {
+      const { width } = properties;
+      return {
+        x: centerX - width / 2,
+      };
+    });
+  };
+
+  const handleAlignRight = () => {
+    onSetProperties((properties) => {
+      const { x, y, width, height, rotationAngle } = properties;
+      let offsetX = 0;
+      if (rotationAngle) {
+        const { width: frameWidth } = calcRotatedObjectPositionAndSize(
+          rotationAngle,
+          x,
+          y,
+          width,
+          height
+        );
+        offsetX = (frameWidth - width) / 2;
+      }
+      return {
+        x: boundRect.endX - width - offsetX,
+      };
+    });
+  };
+
+  const handleAlignTop = () => {
+    onSetProperties((properties) => {
+      const { x, y, width, height, rotationAngle } = properties;
+      let offsetY = 0;
+      if (rotationAngle) {
+        const { height: frameHeight } = calcRotatedObjectPositionAndSize(
+          rotationAngle,
+          x,
+          y,
+          width,
+          height
+        );
+        offsetY = (frameHeight - height) / 2;
+      }
+      return {
+        y: boundRect.startY + offsetY,
+      };
+    });
+  };
+
+  const handleAlignMiddle = () => {
+    const centerY = (boundRect.endY + boundRect.startY) / 2;
+    onSetProperties((properties) => {
+      const { height } = properties;
+      return {
+        y: centerY - height / 2,
+      };
+    });
+  };
+
+  const handleAlignBottom = () => {
+    onSetProperties((properties) => {
+      const { x, y, width, height, rotationAngle } = properties;
+      let offsetY = 0;
+      if (rotationAngle) {
+        const { height: frameHeight } = calcRotatedObjectPositionAndSize(
+          rotationAngle,
+          x,
+          y,
+          width,
+          height
+        );
+        offsetY = (frameHeight - height) / 2;
+      }
+      return {
+        y: boundRect.endY - height - offsetY,
+      };
+    });
+  };
+
+  const handleHorizontalDistribution = () => {
+    updatedSelectedElementsWithFrame.sort(
+      (a, b) => (a.frameX + a.frameWidth) / 2 - (b.frameX + b.frameWidth) / 2
+    );
+    const commonSpaceWidth = dataPixels(
+      (boundRect.width -
+        updatedSelectedElementsWithFrame.reduce(
+          (sum, element) => sum + element.frameWidth,
+          0
+        )) /
+        (updatedSelectedElementsWithFrame.length - 1)
+    );
+    onSetProperties((properties) => {
+      const { id } = properties;
+      const elementIndex = updatedSelectedElementsWithFrame.findIndex(
+        (item) => item.id === id
+      );
+      if (elementIndex === 0 || elementIndex === selectedElements.length - 1) {
+        return {};
+      }
+      const previousElement =
+        updatedSelectedElementsWithFrame[elementIndex - 1];
+      const currentElement = updatedSelectedElementsWithFrame[elementIndex];
+      return {
+        x:
+          previousElement.frameX +
```

Comment:
We have sorted elements here. `updatedSelectedElementsWithFrame`. And also onSetProperties has it's id so we can detect the order of it's object.

Reply:
@dvoytenko sorted order means the selectedElements original orders?

Reply:
Basically each element has common space between them. So element x position should be determined by previous element since every elements has different frameWidth.

Reply:
That makes sense, but I'm not seeing this in your code. `onSetProperties` is not called in determined order and the previous element's `x` and `frameX` could have been changed by this time.

I'm expecting a code like this:

```
const newXs = {};
const offset = 0;
for (let i = 0; i < updatedSelectedElementsWithFrame.length; i++) {
  const {id, x, frameX, frameWidth, width} = updatedSelectedElementsWithFrame[i];
  let newX;
  if (i === 0 || i === updatedSelectedElementsWithFrame.length - 1) {
    newX = x;
  } else {
    newX = offset + commonSpace + (frameWidth - width) / 2;
  }
  newXs[id] = x;
  offset = frameX + (newX - x) + frameWidth;
}
onSetProperties(({id}) => newXs[id])
```

My math could be off. But the most important part is: I don't see how you can do what you're trying to do without the accumulating `offset`. In the current coed `previousElement.frameX/frameWidth` do not give this to you. And to stress: `updatedSelectedElementsWithFrame` and `onSetProperties` are iterating in different order.

Reply:
Now I understand. 🤦‍♂  I assumed that previous element from updatedSelectedElements are already set in correct position. Let me fix this. Thank you

Reply:
Updated the formula!